compiler/noirc_evaluator/src/ssa/opt/basic_conditional.rs

```diff
@@ -0,0 +1,379 @@
+            // including an overhead to take into account the jumps between the blocks.
+            let cost = cost_right.saturating_add(cost_left);
+            if cost < cost / 2 + jump_overhead {
+                if let Some(TerminatorInstruction::JmpIf {
+                    condition: _,
+                    then_destination,
+                    else_destination,
+                    call_stack: _,
+                }) = function.dfg[block].terminator()
+                {
+                    result = Some(BasicConditional {
+                        block_entry: block,
+                        block_then: Some(*then_destination),
+                        block_else: Some(*else_destination),
+                        block_exit: next_left.unwrap(),
+                    });
+                }
+            }
+        } else if left_successors_len == 1 && next_left == Some(right) {
+            // Left branch joins the right branch, it is a if/then statement with no else
+            // I am not sure whether this case can happen, but it is not difficult to handle it
+            let cost = block_cost(left, &function.dfg);
+            if cost < cost / 2 + jump_overhead {
+                if let Some(TerminatorInstruction::JmpIf {
+                    condition: _,
+                    then_destination,
+                    else_destination,
+                    call_stack: _,
+                }) = function.dfg[block].terminator()
+                {
+                    if left == *then_destination {
+                        result = Some(BasicConditional {
+                            block_entry: block,
+                            block_then: Some(left),
+                            block_else: None,
+                            block_exit: right,
+                        });
+                    } else if left == *else_destination {
+                        result = Some(BasicConditional {
+                            block_entry: block,
+                            block_then: None,
+                            block_else: Some(left),
+                            block_exit: right,
+                        });
+                    }
+                }
+            }
+        } else if right_successors_len == 1 && next_right == Some(left) {
+            // Right branch joins the right branch, it is a if/else statement with no then
+            // I am not sure whether this case can happen, but it is not difficult to handle it
+            let cost = block_cost(right, &function.dfg);
+            if cost < cost / 2 + jump_overhead {
+                if let Some(TerminatorInstruction::JmpIf {
+                    condition: _,
+                    then_destination,
+                    else_destination,
+                    call_stack: _,
+                }) = function.dfg[block].terminator()
+                {
+                    if right == *else_destination {
```

Comment:
Same thing as above for the `else if left_successors_len == 1 && next_left == Some(right) {` case

Reply:
Done